cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,326 @@
+        .(ReferenceType)
+        .getBaseType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator->` function for an iterator type.
+ */
+class IteratorFieldOperator extends Operator, TaintFunction {
+  IteratorFieldOperator() {
+    this.hasName("operator->") and
+    this
+        .getACallToThisFunction()
+        .getArgument(0)
+        .getFullyConverted()
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType() instanceof Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator+` function for an iterator type.
+ */
+class IteratorAddOperator extends Operator, TaintFunction {
+  IteratorAddOperator() {
+    this.hasName("operator+") and
+    (
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator or
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator
+    )
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator-` function that takes an iterator as its first argument. This includes
+ * both iterator subtraction and iterator difference overloaded operators.
+ */
+class IteratorSubOperator extends Operator, TaintFunction {
+  IteratorSubOperator() {
+    this.hasName("operator-") and
+    this.getACallToThisFunction().getArgument(0).getType().(PointerType).getBaseType() instanceof
+      Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator+=` or `operator-=` function for an iterator type.
+ */
+class IteratorAssignArithmeticOperator extends Operator, DataFlowFunction, TaintFunction {
+  IteratorAssignArithmeticOperator() {
+    (
+      this.hasName("operator+=") or
+      this.hasName("operator-=")
+    ) and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(1) and
+    output.isParameterDeref(0)
+  }
+}
+
+/**
+ * A non-member `operator[]` function for an iterator type.
```

Comment:
AFAICT from https://en.cppreference.com/w/cpp/language/operator_member_access there is no non-member version of this operator.